### PR TITLE
Expand wildcard names in Zone lookups

### LINKED ISSUE #169: Zone.lookup fails or behaves badly when using wildcards
Hi, i'm using dnsjava in a small authoritative server project (https://github.com/thomaskuh/limbodns) and got 2 issues resolving requests when zone has a wildcard record:

- When lookin up a name handled by a wildcard, the result isn't substituted (Returns *.abc. instead or requestedname.abc.).
- When lookin up a name with type ANY an exception is thrown.

Here's a little test I wrote showing the (in my option) wrong behaviour:
```
public static final void main(String[] args) throws Exception {
	Name nameZone = new Name("example.");
	Zone zone = new Zone(nameZone, new Record[] {
		new SOARecord(nameZone, DClass.IN, 3600L, Name.fromString("nameserver."), new Name("hostmaster",nameZone), 1, 21600L, 7200L, 2160000L, 3600L),
		new NSRecord(nameZone, DClass.IN, 300L, Name.fromString("nameserver.")),
		new ARecord(new Name("test", nameZone), DClass.IN, 3600, Inet4Address.getLocalHost()),
		new ARecord(new Name("*", nameZone), DClass.IN, 3600, Inet4Address.getLocalHost()),
		new TXTRecord(new Name("*", nameZone), DClass.IN, 3600, "sometext")
	});
	
	SetResponse resp = null;
	resp = zone.findRecords(Name.fromString("test.example."), Type.A);
	System.out.println(resp + " -> " + resp.answers());
	// successful -> [{ test.example. 3600 IN A [10.0.0.214] }]
	
	resp = zone.findRecords(Name.fromString("undefined.example."), Type.A);
	System.out.println(resp + " -> " + resp.answers());
	// successful -> [{ *.example. 3600 IN A [10.0.0.214] }]
	// Seems to work but i'd expect:
	// successful -> [{ undefined.example. 3600 IN A [10.0.0.214] }]
	
	resp = zone.findRecords(Name.fromString("undefined.example."), Type.TXT);
	System.out.println(resp + " -> " + resp.answers());
	// successful -> [{ *.example. 3600 IN TXT ["sometext"] }]
	// Seems to work but i'd expect:
	// successful -> [{ undefined.example. 3600 IN TXT ["sometext"] }]
	
	resp = zone.findRecords(Name.fromString("undefined.example."), Type.ANY);
	System.out.println(resp + " -> " + resp.answers());
	// FAILS WITH: java.lang.IllegalArgumentException: oneRRset(ANY) at org.xbill.DNS.Zone.oneRRset(Zone.java:249)
	// Would expect:
	// successful -> [{ *.example. 3600 IN A [10.0.0.214] }, { *.example. 3600 IN TXT ["sometext"] }]
	// or even better:
	// successful -> [{ undefined.example. 3600 IN A [10.0.0.214] }, { undefined.example. 3600 IN TXT ["sometext"] }]
}
```

### PR DESCRIPTION
**Pull request by ibauersachs**
Closes #169

### CODE REVIEW

**Comment by nresare** [comment]
While this certainly works, I would prefer to move towards treating `RRset` as one would an immutable data structure, and instead of using `addRR()` have either a constructor that takes the records or perhaps a more localised solution with a static method that constructs a `RRset` from one or more Record.

I realise we are heavily into opinion and taste territory here, and I am happy to go with whatever solution you have, but since you are asking, I'm happy to share my mine. Feel free to disregard.

I would have done the following:

Created a method to create rrset instances more easily, something along the lines of

```java
private static RRset rrset(Record... records) {
    RRset set = new RRset();
    Arrays.stream(records).forEach(set::addRR);
    return set;
}
```
(for future consideration including into `RRest` as a constructor or static factor method)

Armed with this I would create the `RRset` lists that goes into the first arg of `assertEquals()` inline when I needed them. Possibly, just having the different `Record` instances as `private static final` class fields.

**Comment by nresare** [comment]
My bias is to define the test data that goes into a unit test inline in the actual test most of the time, potentially using an `private static final Type UPPERCASE` constant declaration for values. 

But I do appreciate having something like `assertEquals(singletonList(rrset(LOCALHOST4)), resp.answers())` instead of having the full chain like `assertEquals(1, resp.answers().get(0).rrs().size())` for these reasons:

* The error messages when things are wrong become more descriptive, especially when `toString()` is well written.
* If one want's to be thorough, there are quite a few assert lines that makes sure that the `answers()` method returns exactly the right thing.n
* Sometimes there might be a zero length list return value and the assertEquals might assume that `get(0)` will work and the test failure then becomes an `IndexOutOfBoundsException` which is almost always less helpful than a diff of the `toString()` of the expected and actual values going into the assertion

**Comment by ibauersachs** [comment]
> While this certainly works, I would prefer to move towards treating `RRset` as one would an immutable data structure, and instead of using `addRR()` have either a constructor that takes the records or perhaps a more localised solution with a static method that constructs a `RRset` from one or more Record.

I'd design it like that today, but we can't remove the mutable API. An additional constructor for multiple records is a good idea.

> I realise we are heavily into opinion and taste territory here, and I am happy to go with whatever solution you have, but since you are asking, I'm happy to share my mine. Feel free to disregard.
> 
> I would have done the following:
> 
> Created a method to create rrset instances more easily, something along the lines of
> 
> ```java
> private static RRset rrset(Record... records) {
>     RRset set = new RRset();
>     Arrays.stream(records).forEach(set::addRR);
>     return set;
> }
> ```
> 
> (for future consideration including into `RRest` as a constructor or static factor method)

I prefer the constructor over the static method and added it.

> Armed with this I would create the `RRset` lists that goes into the first arg of `assertEquals()` inline when I needed them. Possibly, just having the different `Record` instances as `private static final` class fields.
>
> My bias is to define the test data that goes into a unit test inline in the actual test most of the time, potentially using an `private static final Type UPPERCASE` constant declaration for values.

The checked exception unfortunately demand a static constructor for that. Almost the same as beforeAll, but at least the fields can be final.

> But I do appreciate having something like `assertEquals(singletonList(rrset(LOCALHOST4)), resp.answers())` instead of having the full chain like `assertEquals(1, resp.answers().get(0).rrs().size())` for these reasons:
> 
> * The error messages when things are wrong become more descriptive, especially when `toString()` is well written.
> * If one want's to be thorough, there are quite a few assert lines that makes sure that the `answers()` method returns exactly the right thing.n
> * Sometimes there might be a zero length list return value and the assertEquals might assume that `get(0)` will work and the test failure then becomes an `IndexOutOfBoundsException` which is almost always less helpful than a diff of the `toString()` of the expected and actual values going into the assertion

Yes, it's perfectly clear why.

File: src/test/java/org/xbill/DNS/ZoneTest.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    assertEquals(
+        listOf(new RRset(A_WILD.withName(testName)), new RRset(TXT_WILD.withName(testName))),
+        resp.answers());
+  }
+
+  private static List<RRset> listOf(RRset... rrsets) {
+    return Stream.of(rrsets).collect(Collectors.toList());
+  }
+
+  private static List<RRset> oneRRset(Record... r) {
```
Comment (nresare) [inline src/test/java/org/xbill/DNS/ZoneTest.java:117]: 👍

File: src/main/java/org/xbill/DNS/RRset.java [hunk 2]
```diff
@@ -35,6 +37,16 @@ public RRset(Record record) {
     addRR(record);
   }
 
+  /** Creates an RRset and sets its contents to the specified record(s) */
+  public RRset(Record... records) {
+    this();
+    if (records != null) {
```
Comment (nresare) [inline src/main/java/org/xbill/DNS/RRset.java:43]: The java vararg design is terrible in that it is possible to pass in a null value of type Record[]. While I think this is fine, I like the `Objects.requireNoNull()` construct instead. If someone explicitly feeds this method a `(Record[])null` they deserve a NullPointerException ;)

**Review by nresare — APPROVED** [review]

**Pull request merged**